### PR TITLE
increased dynamic test suite

### PR DESCRIPTION
**Pull request by BatiGencho**

### CODE REVIEW

File: Makefile [hunk 1]
```diff
@@ -354,15 +354,15 @@ pr:
...
-	cargo run --bin btc-server -- --btc-network "${BITCOIND_NETWORK}" --identifier 0 --address 0.0.0.0:8080 --db "./db1" --min-signers 2 --max-signers 2 --toml ./config.toml  --fee-rate-diff-percentage 30 --bitcoind-url "${BITCOIND_URL}" --bitcoind-user "${BITCOIND_USER}" --bitcoind-pass "${BITCOIND_PWD}" --jwt-secret "${NODE_1_DIR}/jwt.hex" --fall-back-fee-rate-sat-per-vbyte 5
+	cargo run --bin btc-server -- --btc-network "${BITCOIND_NETWORK}" --identifier 0 --address 0.0.0.0:8080 --db "./db1" --min-signers 2 --max-signers 2 --toml ./config.toml  --fee-rate-diff-percentage 30 --bitcoind-url localhost:18443 --bitcoind-user "${BITCOIND_USER}" --bitcoind-pass "${BITCOIND_PWD}" --jwt-secret "${NODE_1_DIR}/jwt.hex" --fall-back-fee-rate-sat-per-vbyte 5
 
 start-btc-server-2:
 	cd ./bin/btc-server && \
-	cargo run --bin btc-server -- --btc-network "${BITCOIND_NETWORK}" --identifier 1 --address 0.0.0.0:8081 --db "./db2" --min-signers 2 --max-signers 2 --toml ./config.toml  --fee-rate-diff-percentage 30 --bitcoind-url "${BITCOIND_URL}" --bitcoind-user "${BITCOIND_USER}" --bitcoind-pass "${BITCOIND_PWD}" --jwt-secret "${NODE_2_DIR}/jwt.hex" --fall-back-fee-rate-sat-per-vbyte 5
```
Comment (arminsabouri) [inline Makefile:365]: Why change this?

File: bin/test-suite/Cargo.toml [hunk 2]
```diff
@@ -35,11 +35,13 @@ reth-btc-wallet = { path = "../../crates/btc-wallet" }
 tempfile.workspace = true
 tracing.workspace = true
 reth-tracing.workspace = true
+reth-ecies.workspace = true
```
Comment (arminsabouri) [inline bin/test-suite/Cargo.toml:38]: do we need ecies as a dependency?

File: bin/test-suite/src/mint_contract_abi.rs [hunk 3]
```diff
@@ -7,12 +7,12 @@ pub use mint_contract::*;
...
     clippy::type_complexity,
     dead_code,
-    non_camel_case_types,
+    non_camel_case_types
 )]
 pub mod mint_contract {
     const _: () = {
         ::core::include_bytes!(
-            "/Users/armins/Dev/botanix/macbeth/bin/test-suite/mint_contract_abi.json",
+            "/home/evgeni/Documents/BOTANIX/Macbeth/bin/test-suite/mint_contract_abi.json",
```
Comment (arminsabouri) [inline bin/test-suite/src/mint_contract_abi.rs:15]: Lets change this to `../mint_contract_abi.json`

File: bin/test-suite/src/suite/consensus/frost/poa_node.rs [hunk 4]
```diff
@@ -284,66 +307,122 @@ pub fn testnet_custom_chain() -> Arc<ChainSpec> {
...
+            None,
+            bitcoin::hash_types::BlockHash::all_zeros(),
+        );
+        let edh = hex::encode(extra_data_header.serialize());
+        let botanix_testnet_config_genesis = BotanixTestnetGenesisConfig { edh: &edh };
+        let rendered_json = botanix_testnet_config_genesis.render().unwrap();
+        let json = serde_json::to_string_pretty(&rendered_json).unwrap();
+        println!("Rendered botanix testnet configuration {:?}", json);
+
+        let botanix_genesis_filepath = Path::new("./").join("botanix_tesnet.json");
```
Comment (arminsabouri) [inline bin/test-suite/src/suite/consensus/frost/poa_node.rs]: Typo
```suggestion
        let botanix_genesis_filepath = Path::new("./").join("botanix_testnet.json");
```

File: bin/test-suite/src/suite/consensus/frost/test_signing.rs [hunk 5]
```diff
@@ -0,0 +1,247 @@
...
+    }
+
+    /*
+    let _finalized = coordinator
+        .finalize_signing(tonic::Request::new(client::FinalizeSigningRequest {
+            signing_session_id: signing_session_id.to_vec(),
+        }))
+        .await
+        .map_err(Error::Request)?;
+    */
```
Comment (arminsabouri) [inline bin/test-suite/src/suite/consensus/frost/test_signing.rs]: Any reason to comment this out?

File: bin/test-suite/src/suite/consensus/frost/test_signing.rs [hunk 6]
```diff
@@ -0,0 +1,247 @@
...
+    }
+
+    /*
+    let _finalized = c3
+        .finalize_signing(tonic::Request::new(client::FinalizeSigningRequest {
+            signing_session_id: signing_session_id.to_vec(),
+        }))
+        .await
+        .map_err(Error::Request)?;
+    */
```
Comment (arminsabouri) [inline bin/test-suite/src/suite/consensus/frost/test_signing.rs]: Samething here

File: bin/test-suite/src/suite/consensus/frost/test_signing.rs (hunk 5 above)
Comment (arminsabouri) [reply bin/test-suite/src/suite/consensus/frost/test_signing.rs]: I think this is the most important part of this test

File: bin/test-suite/src/suite/consensus/pbft/mod.rs [hunk 7]
```diff
@@ -0,0 +1 @@
+
```
Comment (arminsabouri) [inline bin/test-suite/src/suite/consensus/pbft/mod.rs]: Can we delete this?

File: output.txt [hunk 8]
```diff
@@ -0,0 +1,484 @@
+cd ./bin/test-suite && \
```
Comment (arminsabouri) [inline output.txt]: Could delete

**Review by arminsabouri — CHANGES_REQUESTED** [review]
Great work. Just noticed a few things left in that could get deleted

File: bin/test-suite/Cargo.toml (hunk 2 above)
Comment (BatiGencho) [reply bin/test-suite/Cargo.toml:38]: yes, this is needed as we need to derive the peerid from the pub key which reth needs when we are connecting the peers. There is a method inside of that crate that does this for us.

File: bin/test-suite/src/mint_contract_abi.rs (hunk 3 above)
Comment (BatiGencho) [reply bin/test-suite/src/mint_contract_abi.rs:15]: done

File: output.txt (hunk 8 above)
Comment (BatiGencho) [reply output.txt]: done

File: bin/test-suite/src/suite/consensus/frost/poa_node.rs (hunk 4 above)
Comment (BatiGencho) [reply bin/test-suite/src/suite/consensus/frost/poa_node.rs]: done

File: bin/test-suite/src/suite/consensus/pbft/mod.rs (hunk 7 above)
Comment (BatiGencho) [reply bin/test-suite/src/suite/consensus/pbft/mod.rs]: done, i added a .gitkeep file to remind us that pbft is to be added soon there

File: bin/test-suite/src/suite/consensus/frost/test_signing.rs (hunk 5 above)
Comment (BatiGencho) [reply bin/test-suite/src/suite/consensus/frost/test_signing.rs]: The signing part fails in the last step with:

 Failed to broadcast tx: JSON-RPC error: RPC error response: RpcError { code: -25, message: "bad-txns-inputs-missingorspent", data: None }
[2024-04-05T12:21:18Z ERROR btc_server::server] INTERNAL ERROR: Failed to finalize signing: Failed to broadcast tx: JSON-RPC error: RPC error response: RpcError { code: -25, message: "bad-txns-inputs-missingorspent", data: None }

File: bin/test-suite/src/suite/consensus/frost/test_signing.rs (hunk 6 above)
Comment (BatiGencho) [reply bin/test-suite/src/suite/consensus/frost/test_signing.rs]: same as above ,need to investigate a bit further

File: bin/test-suite/src/suite/consensus/frost/test_signing.rs (hunk 5 above)
Comment (arminsabouri) [reply bin/test-suite/src/suite/consensus/frost/test_signing.rs]: This ussually means the signers signed for utxos that don't exist. I can help with this

File: bin/test-suite/src/suite/consensus/frost/test_signing.rs [hunk 9]
```diff
@@ -0,0 +1,247 @@
...
+use client::{BtcServerClient, Round1SigningPackage, Round2SigningPackage};
+use hex::{self, encode as hex_encode};
+use tonic::transport::Channel;
+
+pub async fn test_many_inputs_signing(suite: &ConsensusIntegrationTestSuite) -> Result<(), Error> {
+    let mut eth_addresses: Vec<ethers::core::types::Address> = vec![];
+    for _ in 0..suite.global_context.instances {
+        let eth_address = ethers::core::types::Address::random();
+        eth_addresses.push(eth_address);
+    }
```
Comment (arminsabouri) [inline bin/test-suite/src/suite/consensus/frost/test_signing.rs]: We should have a hardcoded vec of 2. With two eth addresses here. As there should only be 2 inputs being spent here

File: bin/test-suite/src/suite/consensus/frost/test_signing.rs [hunk 10]
```diff
@@ -0,0 +1,247 @@
...
+        assert_eq!(err.code(), tonic::Code::Internal);
+        assert!(err.message().contains("missing key package"));
+    }
+
+    // run the dkg
+    let _ = do_dkg(&mut clients).await?;
+
+    // get the aggregate pk from any of the clients
+    // Here we are signing for a two inputs that are tweaked differently
+    let mut btc_addresses: Vec<Address> = vec![];
```
Comment (arminsabouri) [inline bin/test-suite/src/suite/consensus/frost/test_signing.rs]: Same thing here.

**Comment by stevenroose** [comment]
Just a general comment towards the future, when doing a lot of formatting changes, it would be nice if they could be separated from actual code changes in a separate commit or so.

File: bin/btc-server/src/dkg.rs [hunk 11]
```diff
@@ -86,6 +86,7 @@ impl App {
                 // If we have a max_signers round2 packages we can generate and save the key package
                 let round2_packages = self.db.get_round2_dkg_packages()?;
                 if round2_packages.len() as u16 == self.max_signers - 1 {
+                    // ?
```
Comment (arminsabouri) [inline bin/btc-server/src/dkg.rs]: Did this get resolved?

File: bin/btc-server/src/main.rs [hunk 12]
```diff
@@ -134,12 +135,13 @@ impl App {
         }
         let bitcoind_user = config.bitcoind_user.clone();
         let bitcoind_pass = config.bitcoind_pass.clone();
+        let host = config.bitcoind_url.host_str().unwrap_or_default().to_owned();
+        let port = config.bitcoind_url.port_or_known_default().unwrap_or_default().to_owned();
+        let btcd_url = format!("{}:{}", host, port);
```
Comment (arminsabouri) [inline bin/btc-server/src/main.rs]: Nit: btcd is a alternative bitcoin implementation. can we rename to `bitcoind_url`

File: bin/test-suite/src/suite/consensus/frost/btc_server.rs [hunk 13]
```diff
@@ -80,14 +85,21 @@ pub fn clean_db(tasks: &[SpawnedBtcServer]) {
     }
 }
 
-pub fn spawn_n_btc_servers(n: u16, start_port: u16, config: Config) -> Vec<SpawnedBtcServer> {
+pub fn spawn_n_btc_servers(
+    global_context: Arc<GlobalContext>,
+    start_port: u16,
+) -> Vec<SpawnedBtcServer> {
```
Comment (arminsabouri) [inline bin/test-suite/src/suite/consensus/frost/btc_server.rs:91]: Should # of servers come from a function param. I could see some tests needing a specific set up of nodes. Not a request for a change we could update in the future if need be.

File: bin/test-suite/src/suite/consensus/frost/poa_node.rs [hunk 14]
```diff
@@ -284,66 +323,118 @@ pub fn testnet_custom_chain() -> Arc<ChainSpec> {
...
-    fed_members.insert(1, fed_member_config2);
+    let mut fed_members: HashMap<u16, FederationMemberTestConfig> = HashMap::new();
+
+    let mut members_public_keys: Vec<PublicKey> = vec![];
+    for member_index in 0..global_context.instances {
+        let port = btc_servers
+            .and_then(|servers| servers.iter().nth(member_index as usize).map(|val| val.port))
+            .unwrap();
+
+        let secp: secp256k1::Secp256k1<secp256k1::All> = secp256k1::Secp256k1::new();
```
Comment (arminsabouri) [inline bin/test-suite/src/suite/consensus/frost/poa_node.rs]: Nit: could move this out of the loop

File: bin/test-suite/src/suite/consensus/frost/test_signing.rs [hunk 15]
```diff
@@ -0,0 +1,271 @@
...
+    let finalized = coordinator
+        .finalize_signing(tonic::Request::new(client::FinalizeSigningRequest {
+            signing_session_id: signing_session_id.to_vec(),
+        }))
+        .await;
+
+    assert!(finalized.is_err());
+    let err = finalized.err().unwrap();
+    assert_eq!(err.code(), tonic::Code::Internal);
+    assert!(err.message().contains("bad-txns-inputs-missingorspent"));
```
Comment (arminsabouri) [inline bin/test-suite/src/suite/consensus/frost/test_signing.rs:191]:
```suggestion
    // Test should fail here after `psbt.finalize_mut()`. These inputs don't actually exist on the regtest chain so there is nothing to be spent. In the future we can generate some addresses send funds and use those outpoints for this test.
    assert!(err.message().contains("bad-txns-inputs-missingorspent"));
```

**Review by arminsabouri — COMMENTED** [review]
Looks great! just a few more nits

File: bin/test-suite/src/suite/consensus/frost/poa_node.rs (hunk 14 above)
Comment (stevenroose) [reply bin/test-suite/src/suite/consensus/frost/poa_node.rs]: nit: and the type annotation is not needed ;)

File: bin/btc-server/src/dkg.rs (hunk 11 above)
Comment (BatiGencho) [reply bin/btc-server/src/dkg.rs]: yep, all good here.

File: bin/btc-server/src/main.rs (hunk 12 above)
Comment (BatiGencho) [reply bin/btc-server/src/main.rs]: changed,

File: bin/test-suite/src/suite/consensus/frost/btc_server.rs (hunk 13 above)
Comment (BatiGencho) [reply bin/test-suite/src/suite/consensus/frost/btc_server.rs:91]: the # servers is the max number of signers. I have removed the instance param.

File: bin/test-suite/src/suite/consensus/frost/poa_node.rs (hunk 14 above)
Comment (BatiGencho) [reply bin/test-suite/src/suite/consensus/frost/poa_node.rs]: done

File: bin/test-suite/src/suite/consensus/frost/test_signing.rs (hunk 15 above)
Comment (BatiGencho) [reply bin/test-suite/src/suite/consensus/frost/test_signing.rs:191]: added

**Comment by stevenroose** [comment]
I tried to run it, but didn't really work :/ Various errors. Also the hard-coded paths in .env is not ideal.

```
2024-04-09T16:57:41.030045Z ERROR Test suite panicked PanicInfo { payload: Any { .. }, message: Some(called `Result::unwrap()` on an `Err` value: Bitcoind client could not achieve synced status within 60secs. Exiting...
```

Does the suite check if all the configs in `.env` make sense (like if the dirs exist, if the bitcoind is reachable etc) before running the daemons?

**Review by arminsabouri — APPROVED** [review]

**Pull request merged**